### DISCUSSION TITLE
Proposal: by default, ignore highly generic cosmetic filters

### DISCUSSION THREAD

**Discussion by gorhill** [uBlockOrigin/uBlock-issues#1998] [Ideas]
There are currently 1,137 highly generic cosmetic filters ("HGCF") with default filter lists, which are injected in every page/frame. There is no way to mitigate the impact these have by injecting them conditionally -- as was done with lowly generic cosmetic filters.

These HGCFs are a burden on the browser's rendering engine, for instance consider (picked four at random):

- `##[class*="-slot_ad-placements-"]`
- `##[href*=".jetx.info/"]`
- `##a[href*="//bongacams5.com/track?"]`
- `##[src*="//www.dianomi.com/smartads.epl"]`

You can see all those HGCFs for yourself using the _More_ option in the _Support_ pane, through the _"CFE: Dump"_ button.

There is a lof of HGCFs which try to match using partial attribute value string comparison, and I suspect these are the worst offenders performance-wise on the browser's rendering engine. <https://github.com/uBlockOrigin/uBlock-issues/issues/1983> made the performance hit worse and more visible, but even with the reversion, I investigated further and found that specifically, the HGCFs are having measurable impact on even the most simple web pages.

What I have in mind:

By default, discard all HGCFs which are meant to partially match an attribute value. We now have a _"Report a filter issue"_ button which make it easy for people to report content which should be filtered but is not. Given this, we would build our own filter solutions over time to minimize as much as possible that sort of HGCFs with performance under our control instead of leaving this to 3rd-party filter lists. Where a HGCF is _really_ the only sensible solution, we can use `*##...` to make this explicit as these would not be discarded.

---

Something else to investigate is that whether by dropping HGCFs by default we could make uBO enforce other generic cosmetic filters in Firefox for Android, it could very well be that HGCFs are the only one causing performance issue in Firefox for Android.

**Comment by uBlock-user** [comment]
> ignore highly generic cosmetic filters

Will there be a setting for this ?

**Reply by gorhill** [reply]
Probably _"Ignore highly generic cosmetic filters"_, a descendant of existing _"Ignore generic cosmetic filters"_, checked by default.

I might reverse the settings to _"Accept generic cosmetic filters"_ (enabled by default) and _"Accept highly generic cosmetic filters"_ (disabled by default) to make these less confusing.

**Reply by uBlock-user** [reply]
_Discard generic cosmetic filters_ and _Discard highly generic cosmetic filters_ sounds better imo.

**Comment by u-RraaLL** [comment]
> Given this, we would build our own filter solutions over time to minimize as much as possible that sort of HGCFs with performance under our control instead of leaving this to 3rd-party filter lists.

Does that mean a reversal of this uAssets policy:

> If an issue can be fixed without using extended syntax, i.e. [scriptlet injection](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#scriptlet-injection), [redirect](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#redirect), [procedural cosmetic filters](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#procedural-cosmetic-filters), etc. the issue will, as a rule of thumb, not be fixed by uAssets, because fixing it in EasyList if possible is preferable.

?

**Reply by gorhill** [reply]
If the reported issue is because of an unenforced existing HGCF in EasyList, we will of course fix it in uBO's lists.

**Reply by u-RraaLL** [reply]
Some of the cases reported on uAssets or r/uBlockOrigin are addressed in EL as HGCF.

Would such cases no longer be tagging EL's maintainers or would they be addressed in both EL and uAssets?

**Reply by uBlock-user** [reply]
What will we do about non-default cases -- Fanboy Annoyances list ?

**Reply by gorhill** [reply]
We have _"uBlock filters -- Annoyances" for this. In any case, in the end common sense prevails, so whatever the question is, the answer is what common sense dictates.

**Comment by ghost** [comment]
This should be announced in more places. Maybe at least some of these filters can be avoided. More filter authors should see this. 

Does anchoring to the start or end of an attribute helps at least a bit?

**Reply by gorhill** [reply]
> Does anchoring to the start or end of an attribute helps at least a bit?

Not really. Only when I remove all sorts of `^=`, `*=` and `$=` do I get improved timings -- I could specifically target these and keep other highly generic filters. It could be because of the sheer number of instances of `^=`, there is a lot of them, even more so in _Annoyances_ lists. The only remedy I can think of is to make these specific.

**Comment by ryanbr** [comment]
A few cms's use random chars either added at the start or end of the element, which is when `*=` becomes helpful in EL. But having less `*=` is still more optimal.  Over the next few days, I'll go through EL and trim any generic redundant  `*=` filters.

There isn't many `$=` filters, but trimming/removing `^=` is completely unrealistic. Just too many sites insert random chars into id's and classes.

**Reply by ryanbr** [reply]
https://github.com/easylist/easylist/commit/3a200505
https://github.com/easylist/easylist/commit/6f275386 
as a initial cleanup

**Reply by ryanbr** [reply]
```
pitchfork.com##.jhEvvP.ContentFooterNewsletterContentWrapper-kZJviu
pitchfork.com##.jxhHvg.OutbrainGridColumn-bkXSv

www.starlingbank.com##.CookieModal-module--modalOverlay--22dHh
www.starlingbank.com##.CookieModal-module--modal--1K2OC
```

Example of cosmetics using randomised elements.

**Comment by dportvine** [comment]
When I open many new tabs, the hide rule `##.ya-share2` doesn't work everywhere. But if I add `*##.ya-share2` to my filters, hiding works on all tabs.
You can add `*` for generic filters ##, ###?
Example:  `https://nnmclub.to/` [Imgur](https://imgur.com/NaMwnw7)

<details>

```yaml
uBlock Origin: 1.44.2
Chromium: 105
filterset (summary): 
  network: 98557
  cosmetic: 128386
  scriptlet: 23742
  html: 0
listset (total-discarded, last updated): 
  removed: 
    plowe-0: null
    urlhaus-1: null
  added: 
    https://easylist-downloads.adblockplus.org/bitblock.txt: 31407-27, 2m
    https://easylist-downloads.adblockplus.org/cntblock.txt: 1489-24, 2m
    fanboy-cookiemonster: 40880-285, 2m
    ublock-annoyances: 4569-52, 2m
  default: 
    user-filters: 10-0, never
    RUS-0: 42148-145, 2m
    easylist: 64873-85, 2m
    easyprivacy: 27492-533, 2m
    ublock-abuse: 76-1, 2m
    ublock-badware: 4222-99, 2m
    ublock-filters: 32562-167, 2m
    ublock-privacy: 251-8, 2m
    ublock-quick-fixes: 421-2, 2m
    ublock-unbreak: 1850-44, 2m
filterset (user): [array of 9 redacted]
modifiedUserSettings: 
  advancedUserEnabled: true
  contextMenuEnabled: false
  showIconBadge: false
modifiedHiddenSettings: 
  filterAuthorMode: true
  popupFontSize: 13px
  popupPanelHeightMode: 1
supportStats: 
  allReadyAfter: 753 ms (selfie)
  maxAssetCacheWait: 690 ms
```
</details>

**Reply by u-RraaLL** [reply]
This isn't related to this discussion.

I cannot reproduce on Firefox. Perhaps a Chrome issue.

> You can add `*` for generic filters ##, ###?

Yes, https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#specific-generic
